sql/src/main/java/io/crate/operation/collect/LuceneDocCollector.java

```diff
@@ -200,44 +219,54 @@ public void doCollect() {
         visitorEnabled = fieldsVisitor.required();
         SearchContext.setCurrent(searchContext);
         searchContext.searcher().inStage(ContextIndexSearcher.Stage.MAIN_QUERY);
-        doCollect(null);
+        innerCollect();
+    }
+
+    protected void buildContext(Query query) {
+        if (internalCollectContext == null) {
+            internalCollectContext = new InternalCollectContext(query);
+        }
     }
 
-    private void doCollect(@Nullable OrderedCollectContext collectContext) {
+    protected void searchAndCollect() throws IOException {
+        searchContext.searcher().search(internalCollectContext.query, this);
+    }
+
+    private void innerCollect() {
         try {
-            if (collectContext == null) {
-                Query query = searchContext.query();
-                assert query != null : "query must not be null";
-                if(orderBy != null) {
-                    Sort sort = LuceneSortGenerator.generateLuceneSort(searchContext, orderBy, inputSymbolVisitor);
-                    this.collectContext = new OrderedCollectContext(collectorExpressions, query, orderBy, sort, limit, pageSize);
-                    searchAndCollect(this.collectContext);
-                } else {
-                    searchContext.searcher().search(query, this);
-                }
-            } else {
-                searchAndCollect(collectContext);
-            }
+            Query query = searchContext.query();
+            LOGGER.trace("query {}", query);
+            assert query != null : "query must not be null";
+
+            buildContext(query);
```

Comment:
could this be done in the doCollect call already instead of repeating it? or does the query change somehow?